monkey/common/aws/aws_instance.py

```diff
@@ -1,117 +1,56 @@
-import json
-import logging
-import re
-from dataclasses import dataclass
-from typing import Optional, Tuple
+import threading
 
-import requests
+from .aws_metadata import fetch_aws_instance_metadata
 
-AWS_INSTANCE_METADATA_LOCAL_IP_ADDRESS = "169.254.169.254"
-AWS_LATEST_METADATA_URI_PREFIX = "http://{0}/latest/".format(AWS_INSTANCE_METADATA_LOCAL_IP_ADDRESS)
-ACCOUNT_ID_KEY = "accountId"
+AWS_FETCH_METADATA_TIMEOUT = 10.0  # Seconds
 
-logger = logging.getLogger(__name__)
 
-AWS_TIMEOUT = 2
+class AWSTimeoutError(Exception):
+    """Raised when communications with AWS timeout"""
 
 
-@dataclass
-class AwsInstanceInfo:
-    instance_id: Optional[str] = None
-    region: Optional[str] = None
-    account_id: Optional[str] = None
-
-
-class AwsInstance:
+class AWSInstance:
     """
     Class which gives useful information about the current instance you're on.
     """
 
     def __init__(self):
-        self._is_instance, self._instance_info = AwsInstance._fetch_instance_info()
+        self._instance_id = None
+        self._region = None
+        self._account_id = None
+        self._initialization_complete = threading.Event()
+
+        fetch_thread = threading.Thread(target=self._fetch_aws_instance_metadata, daemon=True)
+        fetch_thread.start()
+
+    def _fetch_aws_instance_metadata(self):
+        metadata = fetch_aws_instance_metadata()
+        self._instance_id = metadata[0]
+        self._region = metadata[1]
+        self._account_id = metadata[2]
```

Comment:
Can we use `NamedTuple` here? I just don't want to check every time what `metadata[0]` can be.

Reply:
We can... That named tuple would, for the most part, serve the same purpose of the `AWSInstance`. It would be somewhat redundant. The `aws_metadata.py` module is only used by `AWSInstance`; It's not exported in the `aws/__init__.py`.

I've made a small change that's simple, but may help scratch the itch.